common/lib/xmodule/xmodule/js/src/problem/edit.coffee

```diff
@@ -189,31 +192,129 @@ class @MarkdownEditingDescriptor extends XModule.Descriptor
+              if (labelmatch) {
+                  hint = hint.replace(labelmatch[0], '').trim();
+                  label = labelmatch[1].trim();
+                  labelassign = ' label="' + label + '"';
+              }
+           }
+           if (detectParens) {
+             if (text.length >= 2 && text[0] == '(' && text[text.length-1] == ')') {
+                 text = text.substring(1, text.length-1)
+                 parens = true;
+              }
+           }
+           return {'nothint': text, 'hint': hint, 'label': label, 'parens': parens, 'labelassign': labelassign};
+      }
+
+
+      // replace selects
+      // [[ a, b, (c) ]]
+      // [[
+      //     a
+      //     b
+      //     (c)
+      //  ]]
+      // <optionresponse>
+      //  <optioninput>
+      //     <option  correct="True">AAA<optionhint  label="Good Job">Yes, multiple choice is the right answer.</optionhint> 
+      // Note: part of the option-response syntax looks like multiple-choice, so it must be processed first.   
+      xml = xml.replace(/\[\[((.|\n)+?)\]\]/g, function(match, group1) {
+          // decide if this is old style or new style
+          if (match.indexOf('\n') == -1) {  // OLD style, [[ .... ]]  on one line
+              var options = group1.split(/\,\s*/g);
+              var optiontag = '  <optioninput options="(';
+              for (i = 0; i < options.length; i += 1) {
+                  optiontag += "'" + options[i].replace(/(?:^|,)\s*\((.*?)\)\s*(?:$|,)/g, '$1') + "'" + (i < options.length -1 ? ',' : '');
+              }
+              optiontag += ')" correct="';
+              var correct = /(?:^|,)\s*\((.*?)\)\s*(?:$|,)/g.exec(group1);
+              if (correct) {
+                  optiontag += correct[1];
+              }
+              optiontag += '">';
+              return '\n<optionresponse>\n' + optiontag + '</optioninput>\n</optionresponse>\n\n';
+          }
+
+          // new style  [[ many-lines ]]
+          var lines = group1.split('\n');
+          var optionlines = ''
+          for (i = 0; i < lines.length; i++) {
+              var line = lines[i].trim();
+              if (line.length > 0) {
+                  var textHint = extractHint(line, true);
+                  var correctstr = ' correct="' + (textHint.parens?'True':'False') + '"';
+                  var hintstr = '';
+                  if (textHint.hint) {
+                      var label = textHint.label;
+                      if (label) {
+                          label = ' label="' + label + '"';
+                      }
+                      hintstr = ' <optionhint' + label + '>' + textHint.hint + '</optionhint>';
+                  }
+                  optionlines += '    <option' + correctstr + '>' + textHint.nothint + hintstr + '</option>\n'
+               }
+           }
+          return '\n<optionresponse>\n  <optioninput>\n' + optionlines + '  </optioninput>\n</optionresponse>\n\n';
```

Comment:
Rather than building up a string using string concatenation, this seems like a good candidate for a Javascript template. Have you checked out [_.template()](http://underscorejs.org/#template)?